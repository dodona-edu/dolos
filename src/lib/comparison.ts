import { HashFilter } from "./hashFilter";
import { NoFilter } from "./noFilter";
import { Tokenizer } from "./tokenizer";
import { WinnowFilter } from "./winnowFilter";

export type Matches<Location> = Map<string, Array<[Location, Location]>>;

/**
 * @param filterFragmentsByPercentage Defines if the fragment should be filtered by percentage or by an absolute value.
 * @param maxFragment The maximum fragment. How this will be used depends on the value of [[filterFragmentByPercentage]]. //TODO better wording needed
 * If it is used as a percentage then the number should be between 0 and 1. If you want to use it as an absolute value
 * then a number between 0 and the amount of files given would be the most useful.
 */
export interface ComparisonFilterOptions {
  maxFragment: number;
  filterFragmentByPercentage: boolean;
}

/**
 * @param hashFilter An optional HashFilter to filter the hashes returned by
 * the rolling hash function.
 * @param noFilter A NoFilter used to generate hashes for blacklisted files.
 */
export interface ComparisonOptions<Location> {
  tokenizer: Tokenizer<Location>;
  hashFilter?: HashFilter;
  noFilter?: NoFilter;
  filterOptions?: ComparisonFilterOptions;
}

export class Comparison<Location> {
  private readonly defaultK: number = 50;
  private readonly defaultW: number = 40;
  // Maps a hash to an array of typles that are composed of the file with the same hash and the location in that file
  // where that hash is located.
  private readonly index: Map<number, Array<[string, Location]>> = new Map();
  private readonly filteredHashSet: Set<number> = new Set();
  private readonly tokenizer: Tokenizer<Location>;
  private readonly hashFilter: HashFilter;
  private readonly noFilter: NoFilter;
  private readonly fragmentFilterOptions: ComparisonFilterOptions | undefined;
  private fileCount: number = 0;

  /**
   * Creates a Comparison object with a given Tokenizer and optional HashFilter.
   * If no HashFilter is given, a new WinnowFilter with k-mer length 50 and windows
   * size 40 will be used.
   *
   * After creation, first add files to the index which can then be queried.
   *
   * @param tokenizer A tokenizer for the correct programming language
   * @param fragmentOptions The options used to filter based on the fragment count. For a more detailed explanation see
   * [[ComparisonOptions]]. If this options is not used then no filtering will occur.
   */
  constructor({ tokenizer, hashFilter, noFilter, filterOptions }: ComparisonOptions<Location>) {
    this.tokenizer = tokenizer;
    this.hashFilter = hashFilter || new WinnowFilter(this.defaultK, this.defaultW);
    this.noFilter = noFilter || new NoFilter(this.defaultK);
    this.fragmentFilterOptions = filterOptions;
  }

  /**
   * Adds a list of files to the comparison index.
   *
   * @param files A list of filenames
   */
  public addFiles(files: string[]): Promise<void[]> {
    // This promise will reject if one of the underlying promises reject
    // which is not what we want. If one file is missing, the others should
    // still be added. In the future, Promise.all can be replaced by
    // Promise.allSettled, but for now the error handling makes sure they
    // are all resolved correctly.
    return Promise.all(files.map(file => this.addFile(file)));
  }

  /**
   * Add a file to the comparison index.
   *
   * @param file The file name of the file to add
   */
  public async addFile(file: string): Promise<void> {
    this.fileCount += 1;
    try {
      const [ast, mapping] = await this.tokenizer.tokenizeFileWithMapping(file);
      for await (const { hash, location } of this.hashFilter.hashesFromString(ast)) {
        // hash and the corresponding line number //
        const match: [string, Location] = [file, mapping[location]];
        const matches = this.index.get(hash);
        if (matches) {
          matches.push(match);
        } else {
          this.index.set(hash, [match]);
        }
      }
    } catch (error) {
      console.error(`There was a problem parsing ${file}. ${error}`);
      return; // this makes sure the promise resolves instead of rejects
    }
  }

  /**
   * Compare a list of files to the index. A map will be returned containing for each
   * file in the input a list of matching files. This list of matching files is also
   * represented as a map, mapping the filename to the position of the match in both files.
   *
   * @param files A list of filenames to query
   * @param hashFilter An optional HashFilter. By default the HashFilter of the Comparison
   * object will be used.
   */
  public async compareFiles(
    files: string[],
    hashFilter = this.hashFilter,
  ): Promise<Map<string, Matches<Location>>> {
    const matchingFiles: Map<string, Matches<Location>> = new Map();
    for (const file of files) {
      matchingFiles.set(file, await this.compareFile(file, hashFilter));
    }
    return matchingFiles;
  }

  /**
   * Compare a file to the index. A map will be returned containing the filename
   * of the matching file, along with a list of matching position between the two files.
   *
   * @param file The file to query
   * @param hashFilter An optional HashFilter. By default the HashFilter of the Comparison
   * object will be used.
   */
  public async compareFile(file: string, hashFilter = this.hashFilter): Promise<Matches<Location>> {
    // mapping file names to line number pairs (other file, this file)
    const matchingFiles: Matches<Location> = new Map();
    const [ast, mapping] = await this.tokenizer.tokenizeFileWithMapping(file);
    for await (const { hash, location } of hashFilter.hashesFromString(ast)) {
      if (this.filteredHashSet.has(hash)) {
        continue;
      }
      const matches = this.index.get(hash);
      if (!matches || !this.filterByFragmentCount(matches.length)) {
        continue;
      }

      for (const [fileName, lineNumber] of matches) {
        // add the match if the match is not with the file and if the file comes first when alphabetically sorted.
        // This is done to avoid the duplicates in the following case: when file A matches with file B, file B will
        // also match with file A.
        if (fileName === file) {
          continue;
        }
        const match: [Location, Location] = [lineNumber, mapping[location]];
        const lines = matchingFiles.get(fileName);
        if (lines) {
          lines.push(match);
        } else {
          matchingFiles.set(fileName, [match]);
        }
      }
    }
    return matchingFiles;
  }

  /**
   * Add a file to the filter index.
   * @param file The file name of the file you want to add.
   */
  public async addFileToFilterList(file: string): Promise<void> {
    try {
      const [ast] = await this.tokenizer.tokenizeFileWithMapping(file);
      for await (const { hash } of this.noFilter.hashesFromString(ast)) {
        this.filteredHashSet.add(hash);
      }
    } catch (error) {
      console.error(`There was a problem parsing ${file}. ${error}`);
      return; // this makes sure the promise resolves instead of rejects
    }
  }

  /**
   *
   * @param fragmentCount The amount of times a certain code fragment has appeared
   * @returns true if is value is acceptable under the options given in the contructor or when the options are not
   * defined.
   */
  private filterByFragmentCount(fragmentCount: number): boolean {
    if (!this.fragmentFilterOptions) {
      return true;
    } else if (this.fragmentFilterOptions.filterFragmentByPercentage) {
      return fragmentCount / this.fileCount <= this.fragmentFilterOptions.maxFragment;
    } else {
      return fragmentCount <= this.fragmentFilterOptions.maxFragment;
    }
  }
}
