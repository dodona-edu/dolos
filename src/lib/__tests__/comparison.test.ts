import { Tokenizer } from "../tokenizer";
import { CodeTokenizer } from "../codeTokenizer";
import { Comparison, Matches } from "../comparison";
const files: string[] = [
    "samples/js/sample.js",
    "samples/js/another_copied_function.js",
    "samples/js/copied_function.js",
];

test("all files no filter test", async () => {
    const tokenizer: Tokenizer<number> = new CodeTokenizer("javascript");
    const comparison: Comparison<number> = new Comparison(
        tokenizer, {filterHashByPercentage: undefined}
    )
    await comparison.addFiles(files);

    const results: Map<string, Matches<number>> = await comparison.compareFiles(files);
    expect(results).toMatchSnapshot();
});

test("all files basefile test", async () => {
    const tokenizer: Tokenizer<number> = new CodeTokenizer("javascript");
    const comparison: Comparison<number> = new Comparison(
        tokenizer, {filterHashByPercentage: undefined}
    )
    await comparison.addFiles(files.filter((_, index) => index !== 1));
    await comparison.addFileToFilterList(files[1]);

    const results: Map<string, Matches<number>> = await comparison.compareFiles(files);
    expect(results).toMatchSnapshot();
});

test("all files max hash count", async () => {
    const tokenizer: Tokenizer<number> = new CodeTokenizer("javascript");
    const comparison: Comparison<number> = new Comparison(
        tokenizer, {filterHashByPercentage: false, maxHash: 4}
    )
    await comparison.addFiles(files);

    const results: Map<string, Matches<number>> = await comparison.compareFiles(files);
    expect(results).toMatchSnapshot();
});

test("all files max hash percentage", async () => {
    const tokenizer: Tokenizer<number> = new CodeTokenizer("javascript");
    const comparison: Comparison<number> = new Comparison(
        tokenizer, {filterHashByPercentage: true, maxHash: 0.4}
    )
    await comparison.addFiles(files);

    const results: Map<string, Matches<number>> = await comparison.compareFiles(files);
    expect(results).toMatchSnapshot();
});