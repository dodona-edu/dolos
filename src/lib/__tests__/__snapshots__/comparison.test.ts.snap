// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all files basefile test 1`] = `
Map {
  "samples/js/sample.js" => Map {
    "samples/js/copied_function.js" => Array [
      Array [
        3,
        14,
      ],
      Array [
        3,
        14,
      ],
      Array [
        4,
        15,
      ],
      Array [
        4,
        15,
      ],
      Array [
        4,
        15,
      ],
      Array [
        5,
        16,
      ],
      Array [
        5,
        16,
      ],
      Array [
        5,
        16,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
    ],
  },
  "samples/js/another_copied_function.js" => Map {},
  "samples/js/copied_function.js" => Map {
    "samples/js/sample.js" => Array [
      Array [
        14,
        3,
      ],
      Array [
        14,
        3,
      ],
      Array [
        15,
        4,
      ],
      Array [
        15,
        4,
      ],
      Array [
        15,
        4,
      ],
      Array [
        16,
        5,
      ],
      Array [
        16,
        5,
      ],
      Array [
        16,
        5,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
    ],
  },
}
`;

exports[`all files max hash count 1`] = `
Map {
  "samples/js/sample.js" => Map {
    "samples/js/another_copied_function.js" => Array [
      Array [
        0,
        2,
      ],
      Array [
        0,
        2,
      ],
      Array [
        3,
        3,
      ],
      Array [
        0,
        11,
      ],
      Array [
        0,
        11,
      ],
      Array [
        0,
        11,
      ],
      Array [
        0,
        11,
      ],
      Array [
        0,
        11,
      ],
      Array [
        0,
        11,
      ],
      Array [
        0,
        11,
      ],
      Array [
        0,
        11,
      ],
      Array [
        4,
        11,
      ],
      Array [
        0,
        11,
      ],
      Array [
        0,
        11,
      ],
      Array [
        0,
        25,
      ],
      Array [
        0,
        27,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        2,
        35,
      ],
      Array [
        2,
        43,
      ],
      Array [
        2,
        51,
      ],
      Array [
        3,
        52,
      ],
      Array [
        3,
        52,
      ],
      Array [
        3,
        52,
      ],
      Array [
        0,
        53,
      ],
      Array [
        4,
        53,
      ],
      Array [
        4,
        53,
      ],
      Array [
        4,
        53,
      ],
      Array [
        4,
        53,
      ],
      Array [
        4,
        53,
      ],
      Array [
        4,
        53,
      ],
      Array [
        4,
        53,
      ],
      Array [
        4,
        53,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
    ],
    "samples/js/copied_function.js" => Array [
      Array [
        3,
        14,
      ],
      Array [
        3,
        14,
      ],
      Array [
        4,
        15,
      ],
      Array [
        4,
        15,
      ],
      Array [
        4,
        15,
      ],
      Array [
        5,
        16,
      ],
      Array [
        5,
        16,
      ],
      Array [
        5,
        16,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
    ],
  },
  "samples/js/another_copied_function.js" => Map {
    "samples/js/sample.js" => Array [
      Array [
        2,
        0,
      ],
      Array [
        25,
        0,
      ],
      Array [
        2,
        0,
      ],
      Array [
        11,
        0,
      ],
      Array [
        27,
        0,
      ],
      Array [
        11,
        0,
      ],
      Array [
        11,
        0,
      ],
      Array [
        11,
        0,
      ],
      Array [
        11,
        0,
      ],
      Array [
        11,
        0,
      ],
      Array [
        11,
        0,
      ],
      Array [
        11,
        0,
      ],
      Array [
        53,
        0,
      ],
      Array [
        11,
        0,
      ],
      Array [
        11,
        0,
      ],
      Array [
        35,
        2,
      ],
      Array [
        43,
        2,
      ],
      Array [
        51,
        2,
      ],
      Array [
        52,
        3,
      ],
      Array [
        3,
        3,
      ],
      Array [
        52,
        3,
      ],
      Array [
        52,
        3,
      ],
      Array [
        11,
        4,
      ],
      Array [
        53,
        4,
      ],
      Array [
        53,
        4,
      ],
      Array [
        53,
        4,
      ],
      Array [
        53,
        4,
      ],
      Array [
        53,
        4,
      ],
      Array [
        53,
        4,
      ],
      Array [
        53,
        4,
      ],
      Array [
        53,
        4,
      ],
      Array [
        54,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        54,
        5,
      ],
    ],
  },
  "samples/js/copied_function.js" => Map {
    "samples/js/sample.js" => Array [
      Array [
        14,
        3,
      ],
      Array [
        14,
        3,
      ],
      Array [
        15,
        4,
      ],
      Array [
        15,
        4,
      ],
      Array [
        15,
        4,
      ],
      Array [
        16,
        5,
      ],
      Array [
        16,
        5,
      ],
      Array [
        16,
        5,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
    ],
  },
}
`;

exports[`all files max hash percentage 1`] = `
Map {
  "samples/js/sample.js" => Map {},
  "samples/js/another_copied_function.js" => Map {},
  "samples/js/copied_function.js" => Map {},
}
`;

exports[`all files no filter test 1`] = `
Map {
  "samples/js/sample.js" => Map {
    "samples/js/another_copied_function.js" => Array [
      Array [
        0,
        2,
      ],
      Array [
        0,
        2,
      ],
      Array [
        0,
        2,
      ],
      Array [
        0,
        2,
      ],
      Array [
        0,
        2,
      ],
      Array [
        0,
        2,
      ],
      Array [
        0,
        2,
      ],
      Array [
        3,
        3,
      ],
      Array [
        4,
        3,
      ],
      Array [
        5,
        3,
      ],
      Array [
        5,
        3,
      ],
      Array [
        5,
        7,
      ],
      Array [
        4,
        7,
      ],
      Array [
        5,
        7,
      ],
      Array [
        5,
        7,
      ],
      Array [
        4,
        7,
      ],
      Array [
        5,
        7,
      ],
      Array [
        0,
        11,
      ],
      Array [
        0,
        11,
      ],
      Array [
        0,
        11,
      ],
      Array [
        0,
        11,
      ],
      Array [
        0,
        11,
      ],
      Array [
        0,
        11,
      ],
      Array [
        0,
        11,
      ],
      Array [
        0,
        11,
      ],
      Array [
        0,
        11,
      ],
      Array [
        0,
        11,
      ],
      Array [
        0,
        11,
      ],
      Array [
        0,
        11,
      ],
      Array [
        0,
        11,
      ],
      Array [
        4,
        11,
      ],
      Array [
        0,
        11,
      ],
      Array [
        0,
        11,
      ],
      Array [
        5,
        17,
      ],
      Array [
        0,
        25,
      ],
      Array [
        0,
        25,
      ],
      Array [
        0,
        25,
      ],
      Array [
        0,
        25,
      ],
      Array [
        0,
        25,
      ],
      Array [
        0,
        25,
      ],
      Array [
        0,
        26,
      ],
      Array [
        0,
        26,
      ],
      Array [
        0,
        26,
      ],
      Array [
        0,
        26,
      ],
      Array [
        0,
        26,
      ],
      Array [
        0,
        27,
      ],
      Array [
        0,
        27,
      ],
      Array [
        0,
        27,
      ],
      Array [
        0,
        27,
      ],
      Array [
        0,
        27,
      ],
      Array [
        0,
        27,
      ],
      Array [
        0,
        28,
      ],
      Array [
        0,
        28,
      ],
      Array [
        0,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        4,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        4,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        4,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        4,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        5,
        28,
      ],
      Array [
        2,
        35,
      ],
      Array [
        0,
        36,
      ],
      Array [
        5,
        36,
      ],
      Array [
        4,
        36,
      ],
      Array [
        5,
        36,
      ],
      Array [
        5,
        36,
      ],
      Array [
        2,
        43,
      ],
      Array [
        4,
        44,
      ],
      Array [
        5,
        44,
      ],
      Array [
        5,
        44,
      ],
      Array [
        2,
        51,
      ],
      Array [
        3,
        52,
      ],
      Array [
        3,
        52,
      ],
      Array [
        3,
        52,
      ],
      Array [
        0,
        53,
      ],
      Array [
        4,
        53,
      ],
      Array [
        4,
        53,
      ],
      Array [
        4,
        53,
      ],
      Array [
        5,
        53,
      ],
      Array [
        5,
        53,
      ],
      Array [
        4,
        53,
      ],
      Array [
        4,
        53,
      ],
      Array [
        4,
        53,
      ],
      Array [
        4,
        53,
      ],
      Array [
        5,
        53,
      ],
      Array [
        5,
        53,
      ],
      Array [
        4,
        53,
      ],
      Array [
        5,
        53,
      ],
      Array [
        4,
        53,
      ],
      Array [
        4,
        53,
      ],
      Array [
        4,
        53,
      ],
      Array [
        4,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        4,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        4,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        4,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        4,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
      Array [
        5,
        54,
      ],
    ],
    "samples/js/copied_function.js" => Array [
      Array [
        3,
        14,
      ],
      Array [
        3,
        14,
      ],
      Array [
        4,
        15,
      ],
      Array [
        4,
        15,
      ],
      Array [
        4,
        15,
      ],
      Array [
        5,
        16,
      ],
      Array [
        5,
        16,
      ],
      Array [
        5,
        16,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        17,
      ],
      Array [
        6,
        28,
      ],
      Array [
        6,
        54,
      ],
    ],
  },
  "samples/js/another_copied_function.js" => Map {
    "samples/js/sample.js" => Array [
      Array [
        2,
        0,
      ],
      Array [
        25,
        0,
      ],
      Array [
        2,
        0,
      ],
      Array [
        11,
        0,
      ],
      Array [
        25,
        0,
      ],
      Array [
        26,
        0,
      ],
      Array [
        27,
        0,
      ],
      Array [
        28,
        0,
      ],
      Array [
        2,
        0,
      ],
      Array [
        11,
        0,
      ],
      Array [
        25,
        0,
      ],
      Array [
        26,
        0,
      ],
      Array [
        27,
        0,
      ],
      Array [
        28,
        0,
      ],
      Array [
        2,
        0,
      ],
      Array [
        11,
        0,
      ],
      Array [
        25,
        0,
      ],
      Array [
        26,
        0,
      ],
      Array [
        27,
        0,
      ],
      Array [
        28,
        0,
      ],
      Array [
        2,
        0,
      ],
      Array [
        11,
        0,
      ],
      Array [
        25,
        0,
      ],
      Array [
        26,
        0,
      ],
      Array [
        27,
        0,
      ],
      Array [
        36,
        0,
      ],
      Array [
        2,
        0,
      ],
      Array [
        11,
        0,
      ],
      Array [
        25,
        0,
      ],
      Array [
        26,
        0,
      ],
      Array [
        27,
        0,
      ],
      Array [
        2,
        0,
      ],
      Array [
        11,
        0,
      ],
      Array [
        27,
        0,
      ],
      Array [
        11,
        0,
      ],
      Array [
        11,
        0,
      ],
      Array [
        11,
        0,
      ],
      Array [
        11,
        0,
      ],
      Array [
        11,
        0,
      ],
      Array [
        11,
        0,
      ],
      Array [
        11,
        0,
      ],
      Array [
        53,
        0,
      ],
      Array [
        11,
        0,
      ],
      Array [
        11,
        0,
      ],
      Array [
        35,
        2,
      ],
      Array [
        43,
        2,
      ],
      Array [
        51,
        2,
      ],
      Array [
        52,
        3,
      ],
      Array [
        3,
        3,
      ],
      Array [
        52,
        3,
      ],
      Array [
        52,
        3,
      ],
      Array [
        11,
        4,
      ],
      Array [
        53,
        4,
      ],
      Array [
        53,
        4,
      ],
      Array [
        28,
        4,
      ],
      Array [
        28,
        4,
      ],
      Array [
        53,
        4,
      ],
      Array [
        54,
        4,
      ],
      Array [
        54,
        4,
      ],
      Array [
        53,
        4,
      ],
      Array [
        53,
        4,
      ],
      Array [
        53,
        4,
      ],
      Array [
        3,
        4,
      ],
      Array [
        7,
        4,
      ],
      Array [
        28,
        4,
      ],
      Array [
        36,
        4,
      ],
      Array [
        44,
        4,
      ],
      Array [
        53,
        4,
      ],
      Array [
        54,
        4,
      ],
      Array [
        54,
        4,
      ],
      Array [
        7,
        4,
      ],
      Array [
        28,
        4,
      ],
      Array [
        53,
        4,
      ],
      Array [
        54,
        4,
      ],
      Array [
        53,
        4,
      ],
      Array [
        53,
        4,
      ],
      Array [
        53,
        4,
      ],
      Array [
        3,
        5,
      ],
      Array [
        7,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        36,
        5,
      ],
      Array [
        44,
        5,
      ],
      Array [
        53,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        17,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        7,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        36,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        3,
        5,
      ],
      Array [
        7,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        36,
        5,
      ],
      Array [
        44,
        5,
      ],
      Array [
        53,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        7,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        53,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        53,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        53,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        54,
        5,
      ],
      Array [
        28,
        5,
      ],
      Array [
        54,
        5,
      ],
    ],
    "samples/js/copied_function.js" => Array [
      Array [
        6,
        5,
      ],
    ],
  },
  "samples/js/copied_function.js" => Map {
    "samples/js/sample.js" => Array [
      Array [
        14,
        3,
      ],
      Array [
        14,
        3,
      ],
      Array [
        15,
        4,
      ],
      Array [
        15,
        4,
      ],
      Array [
        15,
        4,
      ],
      Array [
        16,
        5,
      ],
      Array [
        16,
        5,
      ],
      Array [
        16,
        5,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        28,
        6,
      ],
      Array [
        54,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
      Array [
        17,
        6,
      ],
    ],
    "samples/js/another_copied_function.js" => Array [
      Array [
        5,
        6,
      ],
    ],
  },
}
`;
